Keep window on the same display when its client size is changed

vadz opened the pull request:
The really important commit here is 7aed4291e0c10e911039a0e791176d0eb7f98dd6, which fixes an infinite recursion.

The rest are support commits adding/implementing `wxDisplay::GetFromRect()` which might be useful to have in other cases too and is needed here.

cc @MaartenBent

vadz committed:
Add wxDisplay::GetFromRect()

For now this is just the same thing as GetFromPoint(center-of-rectangle)
but it's already more convenient to use and it will also be changed in
the upcoming commits.

vadz committed:
Keep window on the same display in SetClientSize()

It was unexpected that changing the window client size could change the
display the window was on and, even worse, result in further size
changes because changing the display could change the DPI scaling used.
And, worst of all, if the window wxEVT_DPI_CHANGED handler called
SetClientSize(), possibly indirectly such as via wxSizer::Fit(), this
could result in an infinite recursion when moving the window to a high
definition display located to the left of the standard definition one
(the problem didn't happen for the reciprocal display arrangement
because increasing the window size couldn't change the window display
in that case).

So ensure that the window remains on the same display it is on now, by
keeping its center position unchanged, instead of keeping the position
of its top left corner as we did before.

vadz committed:
Reuse GetFromRect() for non-wxMSW GetFromWindow() implementation

Use the new function here instead of duplicating what it does in
non-wxMSW ports using MSW API.

vadz committed:
Provide native implementation of GetFromRect() for wxMSW

Use the Win32 API function directly doing what we need.

vadz committed:
Improve generic implementation of wxDisplay::GetFromRect()

Account for the special case of partly visible rectangle whose center is
off screen, as in this case GetFromRect() would still be expected to
return a valid display and not wxNOT_FOUND.

Also mention this case in the documentation.

vadz committed:
Avoid mis-detecting NULL as part of other identifiers

Tighten the search for NULL in the code style check to avoid flagging
the use of MONITOR_DEFAULTTONULL (Win32 API constant) as violation.

vadz commented on src/msw/window.cpp:
```diff
@@ -2288,8 +2292,33 @@ void wxWindowMSW::DoSetClientSize(int width, int height)
+                // of the window position sufficient for keeping it on the
+                // current display, so keep things simple and preserve the
+                // position of its center horizontally.
+                proposedRect.x = rectWin.left - proposedRect.width/2;
```
You're absolutely right, of course, thanks for noticing this, I've hesitated between writing this in 2 different ways and ended up with a middle version which makes no sense :-(

Fixed now, thanks again!

MaartenBent replied:
Maybe you can also use `proposedRect.CenterIn(wxRectFromRECT(rectWin), wxHORIZONTAL)`? I think this does the same.

vadz replied:
It would need to be `proposedRect = proposedRect.CenterIn(...)`, but I've added `MakeCenteredIn()` to make this more clear now -- what do you think?

MaartenBent replied:
Looks good!

vadz committed:
Get rid of CppUnit boilerplate in wxRect unit test

No real changes, just clean up the test.

vadz committed:
Add wxRect::MakeCenteredIn() modifying the rectangle in place

This can be more convenient than calling CenterIn().

vadz committed:
Use wxRect::MakeCenteredIn() in DoSetClientSize()

No real changes, just make the code more clear by adding the new helper
function.
